Changing order of callback and dismissing the view controller

jennyf19 opened the pull request:
Addressing issue #1016 
Customer reports issues with navigation stack when authentication is cancelled and we perform the callback first and then `DismissViewController`. It makes sense to shut down the UI first, especially for customers using more complex view controllers

Manually tested:
> Xamarin iOS test app

jennyf19 committed:
Changing order of callback and dismissing the view controller

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs:
```diff
@@ -152,8 +150,8 @@ public override void ViewDidLoad()
         {
-            callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null));
             this.DismissViewController(true, null);
+            callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null));
```
isn't the override on the DismissViewController a callback to a method which will be ensure to be completed after dismissal of the viewcontroller? (see below) - I think the right thing to do is to :

this.DismissViewController(true, new AuthorizationResult( ...)

(wrapped into an action perhaps via the callbackMethod)

jennyf19 replied:
I have a few questions about this for you...I'll swing by tomorrow. 🐈

henrik-me reviewed (changes requested):
:clock1:

gameleon-dev commented:
The `DismissViewController` method is not synchronous. Which means there is no guarantee the view controller has been properly removed from the hierarchy even if you switch the method calls in the code.

The proper way to handle this would be to either use the completion handler:

```
DismissViewController(true, () =>
{
    callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null));
});
```

Or use the Async version of the dismiss method:

```
await this.DismissViewControllerAsync(true);
callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null));
```

Both these methods will ensure that the viewcontroller has been removed from the view hierachy.

gameleon-dev commented:
Looking at the controller it seems there are a two more snippets where this might be an issue:

https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/82565905ca0213e7963a275c40382a8560f9d141/adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs#L95-L101

and 

https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/62a82c78b2a8504065f342309f8015dde8b9eccc/adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs#L128-L129

henrik-me commented:
Thanks for sharing the info @gameleon-dev. We have had a few internal discussions on this. We'll take a look at it. @jennyf19 @jmprieur

gameleon-dev commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs:
```diff
@@ -152,8 +150,8 @@ public override void ViewDidLoad()
         {
-            callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null));
             this.DismissViewController(true, null);
+            callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null));
```
I would propose to change this line to:

`this.DismissViewController(true, ()=> callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null)));`

Moving the callback method to the completionHandler would ensure it will not fire until the controller is fully dismissed and removed from the view hierarchy.

The same would apply to the following code snippets:

https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/40e622f5aa279bf7c5852cdca8bd572580039afa/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs#L96-L97

https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/40e622f5aa279bf7c5852cdca8bd572580039afa/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs#L126-L127

These would have to be changed to 

`
                    this.DismissViewController(true, ()=>
                        callbackMethod(new AuthorizationResult(AuthorizationStatus.Success, request.Url.ToString())));`

and 

`
                    this.DismissViewController(true, ()=> callbackMethod(result));`

respectively.

gameleon-dev commented:
Apologies for the delay on this one. I have made the proposed changes in #1095

leonluc-dev committed:
Move callback calls to completionHandler of DismissViewController (#1095)

This will ensure that the view controller is fully dismissed and removed from the view hierarchy before the callback method fires.

jennyf19 committed:
Merge remote-tracking branch 'origin/adalv3/dev' into jennyf/iOSViewController

jennyf19 commented:
@henrik-me - I pulled in @gameleon-dev 's changes and updated the branch. Manually tested w/iOS simulator.

---
In reply to: [383859682](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1044#issuecomment-383859682) [](ancestors = 383859682)

henrik-me commented:
Thanks, @jennyf19 if I recall correctly you made some similar changes to the dev branch and thus we should be good to go. Approving.
